src/PackagePrivate/DatabaseTermIdsAcquirer.php

```diff
@@ -0,0 +1,420 @@
+<?php
+
+namespace Wikibase\TermStore\MediaWiki\PackagePrivate;
+
+use Wikimedia\Rdbms\IDatabase;
+
+class DatabaseTermIdsAcquirer implements TermIdsAcquirer {
+
+	const TABLE_TEXT = 'wbt_text';
+	const TABLE_TEXT_IN_LANG = 'wbt_text_in_lang';
+	const TABLE_TERM_IN_LANG = 'wbt_term_in_lang';
+
+	/**
+	 * @var Database $dbw
+	 */
+	private $dbw;
+
+	/**
+	 * @var Database $dbr
+	 */
+	private $dbr;
+
+	public function __construct(
+		IDatabase $dbw,
+		IDatabase $dbr,
+		TypeIdsAcquirer $typeIdsAcquirer
+	) {
+		$this->dbw = $dbw;
+		$this->dbr = $dbr;
+		$this->typeIdsAcquirer = $typeIdsAcquirer;
+	}
+
+	public function acquireTermIds( array $termsArray ): array {
+		$termsArray = $this->mapToTextIds( $termsArray );
+		$termsArray = $this->mapToTextInLangIds( $termsArray );
+		$termsArray = $this->mapToTypeIds( $termsArray );
+		return $this->mapToTermInLangIds( $termsArray );
+	}
+
+	/**
+	 * replace root keys containing type names in termsArray
+	 * with their respective ids in wbt_type table
+	 *
+	 * @param array $termsArray terms per type per language:
+	 *	[
+	 *		'type1' => [ ... ],
+	 *		'type2' => [ ... ],
+	 *		...
+	 *	]
+	 *
+	 * @return array
+	 *	[
+	 *		<typeId1> => [ ... ],
+	 *		<typeId2> => [ ... ],
+	 *		...
+	 *	]
+	 */
+	private function mapToTypeIds( array $termsArray ) {
+		$typeIds = $this->typeIdsAcquirer->acquireTypeIds( array_keys( $termsArray ) );
+
+		$termsArrayByTypeId = [];
+		foreach ( $typeIds as $type => $typeId ) {
+			$termsArrayByTypeId[$typeId] = $termsArray[$type];
+		}
+		return $termsArrayByTypeId;
+	}
+
+	/**
+	 * replace text at termsArray leaves with their ids in wbt_text table
+	 * and return resulting array
+	 *
+	 * @param array $termsArray terms per type per language:
+	 *	[
+	 *		'type' => [
+	 *			[ 'language' => 'term' | [ 'term1', 'term2', ... ] ], ...
+	 *		], ...
+	 *	]
+	 *
+	 * @return array
+	 *	[
+	 *		'type' => [
+	 *			[ 'language' => [ <textId1>, <textId2>, ... ] ], ...
```

Comment:
Given `[ 'type' => [ 'language' => 'term' ] ]`, will this return `[ 'type' => [ 'language' => 1 ] ]` or `[ 'type' => [ 'language' => [ 1 ] ] ]`? That is, if the input has a single term instead of a list, is that preserved in the output?

Reply:
it will not be preserved iirc, I convert them to arrays in the output even for single elements